File: coding_examples/snake_game.py
# ============================================================================
#  Python Snake Game — Terminal Edition
#  Same logic as snake_game.nova for direct comparison.
# ============================================================================

from dataclasses import dataclass
from enum import Enum, auto
from typing import List, Optional
import math

# --- Data types ---

@dataclass
class Point:
    x: int
    y: int

    def eq(self, other: 'Point') -> bool:
        return self.x == other.x and self.y == other.y

    def to_string(self) -> str:
        return f"({self.x},{self.y})"

class Direction(Enum):
    Up = auto()
    Down = auto()
    Left = auto()
    Right = auto()

@dataclass
class GameState:
    snake: List[Point]
    dir: Direction
    food: Point
    score: int
    width: int
    height: int
    alive: bool
    ticks: int

# --- Direction helpers ---

def direction_dx(d: Direction) -> int:
    if d == Direction.Left: return -1
    if d == Direction.Right: return 1
    return 0

def direction_dy(d: Direction) -> int:
    if d == Direction.Up: return -1
    if d == Direction.Down: return 1
    return 0

def opposite_dir(d1: Direction, d2: Direction) -> bool:
    return (direction_dx(d1) + direction_dx(d2) == 0 and
            direction_dy(d1) + direction_dy(d2) == 0)

# --- Pseudo-random number generator (LCG) ---

_rng_state = 42

def rand_next() -> int:
    global _rng_state
    _rng_state = (_rng_state * 1103515245 + 12345) % 2147483648
    return _rng_state

def rand_range(lo: int, hi: int) -> int:
    r = rand_next()
    if r < 0: r = -r
    return lo + (r % (hi - lo))

# --- Point-in-list check ---

def list_contains_point(lst: List[Point], pt: Point) -> bool:
    for p in lst:
        if p.x == pt.x and p.y == pt.y:
            return True
    return False

# --- Spawn food at random empty cell ---

def spawn_food(snake: List[Point], w: int, h: int) -> Point:
    for _ in range(1000):
        fx = rand_range(0, w)
        fy = rand_range(0, h)
        pt = Point(fx, fy)
        if not list_contains_point(snake, pt):
            return pt
    return Point(0, 0)

# --- Create initial game ---

def new_game(w: int, h: int) -> GameState:
    cx = w // 2
    cy = h // 2
    snake = [
        Point(cx, cy),
        Point(cx - 1, cy),
        Point(cx - 2, cy)
    ]
    food = spawn_food(snake, w, h)
    return GameState(
        snake=snake,
        dir=Direction.Right,
        food=food,
        score=0,
        width=w,
        height=h,
        alive=True,
        ticks=0
    )

# --- One game tick: move, eat, collide ---

def tick(game: GameState, new_dir: Direction) -> GameState:
    if not game.alive:
        return game

    dir = game.dir
    if not opposite_dir(dir, new_dir):
        dir = new_dir

    head = game.snake[0]
    nx = head.x + direction_dx(dir)
    ny = head.y + direction_dy(dir)

    # Wall collision
    if nx < 0 or nx >= game.width or ny < 0 or ny >= game.height:
        return GameState(
            snake=game.snake, dir=dir, food=game.food,
            score=game.score, width=game.width, height=game.height,
            alive=False, ticks=game.ticks + 1
        )

    new_head = Point(nx, ny)

    # Self collision
    if list_contains_point(game.snake, new_head):
        return GameState(
            snake=game.snake, dir=dir, food=game.food,
            score=game.score, width=game.width, height=game.height,
            alive=False, ticks=game.ticks + 1
        )

    # Move: build new snake with head prepended
    ate = new_head.x == game.food.x and new_head.y == game.food.y
    limit = len(game.snake) if ate else len(game.snake) - 1
    new_snake = [new_head] + game.snake[:limit]

    new_score = game.score
    new_food = game.food
    if ate:
        new_score += 10
        new_food = spawn_food(new_snake, game.width, game.height)

    return GameState(
        snake=new_snake, dir=dir, food=new_food,
        score=new_score, width=game.width, height=game.height,
        alive=True, ticks=game.ticks + 1
    )

# --- Render game board to terminal ---

def render(game: GameState) -> None:
    top = "+" + "--" * game.width + "+"
    print(top)

    for y in range(game.height):
        row = "|"
        for x in range(game.width):
            pt = Point(x, y)
            if pt.x == game.snake[0].x and pt.y == game.snake[0].y:
                row += "@@"
            elif list_contains_point(game.snake, pt):
                row += "##"
            elif pt.x == game.food.x and pt.y == game.food.y:
                row += "<>"
            else:
                row += "  "
        row += "|"
        print(row)

    print(top)
    print(f"  Score: {game.score}  |  Length: {len(game.snake)}  |  Tick: {game.ticks}")

# --- Auto-play: simple AI that chases food ---

def auto_dir(game: GameState) -> Direction:
    head = game.snake[0]
    dx = game.food.x - head.x
    dy = game.food.y - head.y

    if abs(dx) >= abs(dy):
        if dx > 0: return Direction.Right
        if dx < 0: return Direction.Left
    if dy > 0: return Direction.Down
    if dy < 0: return Direction.Up
    return game.dir

# --- Main ---

def main():
    print("========================================")
    print("  Python Snake Game — Auto-Play Demo")
    print("  For comparison with NOVA version.")
    print("========================================")
    print()

    game = new_game(20, 15)
    render(game)

    for step in range(60):
        if not game.alive:
            break
        d = auto_dir(game)
        game = tick(game, d)

        if step % 10 == 9:
            print()
            print(f"--- After {step + 1} moves ---")
            render(game)

    print()
    if game.alive:
        print(f"Game still running after {game.ticks} ticks!")
    else:
        print(f"Game over at tick {game.ticks}!")
    print(f"Final score: {game.score}")
    print(f"Final length: {len(game.snake)}")

if __name__ == "__main__":
    main()

File: coding_examples/snake_game.cpp
// ============================================================================
//  C++ Snake Game — REAL Interactive Terminal Game (Windows)
//  Use arrow keys to play. Press Q to quit.
// ============================================================================

#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <conio.h>
#include <windows.h>

struct Point { int x, y; };

enum Direction { UP, DOWN, LEFT, RIGHT };

struct Game {
    std::vector<Point> snake;
    Direction dir;
    Point food;
    int score;
    int width;
    int height;
    bool alive;
};

void set_cursor(int x, int y) {
    COORD pos = {(SHORT)x, (SHORT)y};
    SetConsoleCursorPosition(GetStdHandle(STD_OUTPUT_HANDLE), pos);
}

void hide_cursor() {
    CONSOLE_CURSOR_INFO ci = {1, FALSE};
    SetConsoleCursorInfo(GetStdHandle(STD_OUTPUT_HANDLE), &ci);
}

void show_cursor() {
    CONSOLE_CURSOR_INFO ci = {1, TRUE};
    SetConsoleCursorInfo(GetStdHandle(STD_OUTPUT_HANDLE), &ci);
}

bool contains(const std::vector<Point>& pts, int x, int y) {
    for (const auto& p : pts)
        if (p.x == x && p.y == y) return true;
    return false;
}

void spawn_food(Game& g) {
    do {
        g.food.x = rand() % g.width;
        g.food.y = rand() % g.height;
    } while (contains(g.snake, g.food.x, g.food.y));
}

Game new_game(int w, int h) {
    Game g;
    g.width = w;
    g.height = h;
    g.dir = RIGHT;
    g.score = 0;
    g.alive = true;
    int cx = w / 2, cy = h / 2;
    g.snake = {{cx, cy}, {cx-1, cy}, {cx-2, cy}};
    spawn_food(g);
    return g;
}

void render(const Game& g) {
    set_cursor(0, 0);

    // Title
    std::cout << "  SNAKE GAME  |  Score: " << g.score
              << "  |  Length: " << g.snake.size()
              << "       " << std::endl;
    std::cout << "  Arrow keys = move  |  Q = quit" << std::endl;

    // Top border
    std::cout << "+";
    for (int i = 0; i < g.width; i++) std::cout << "--";
    std::cout << "+" << std::endl;

    // Board
    for (int y = 0; y < g.height; y++) {
        std::cout << "|";
        for (int x = 0; x < g.width; x++) {
            if (x == g.snake[0].x && y == g.snake[0].y)
                std::cout << "@@";
            else if (contains(g.snake, x, y))
                std::cout << "##";
            else if (x == g.food.x && y == g.food.y)
                std::cout << "<>";
            else
                std::cout << "  ";
        }
        std::cout << "|" << std::endl;
    }

    // Bottom border
    std::cout << "+";
    for (int i = 0; i < g.width; i++) std::cout << "--";
    std::cout << "+" << std::endl;
}

void tick(Game& g) {
    Point head = g.snake[0];
    switch (g.dir) {
        case UP:    head.y--; break;
        case DOWN:  head.y++; break;
        case LEFT:  head.x--; break;
        case RIGHT: head.x++; break;
    }

    // Wall collision
    if (head.x < 0 || head.x >= g.width || head.y < 0 || head.y >= g.height) {
        g.alive = false;
        return;
    }

    // Self collision
    if (contains(g.snake, head.x, head.y)) {
        g.alive = false;
        return;
    }

    g.snake.insert(g.snake.begin(), head);

    // Eat food
    if (head.x == g.food.x && head.y == g.food.y) {
        g.score += 10;
        spawn_food(g);
    } else {
        g.snake.pop_back();
    }
}

int main() {
    srand((unsigned)time(nullptr));
    system("cls");
    hide_cursor();

    Game game = new_game(20, 15);

    while (game.alive) {
        render(game);

        // Non-blocking input
        int wait = 0;
        while (wait < 150) {
            if (_kbhit()) {
                int ch = _getch();
                if (ch == 'q' || ch == 'Q') {
                    game.alive = false;
                    break;
                }
                if (ch == 0 || ch == 224) {
                    ch = _getch();
                    switch (ch) {
                        case 72: if (game.dir != DOWN)  game.dir = UP;    break;
                        case 80: if (game.dir != UP)    game.dir = DOWN;  break;
                        case 75: if (game.dir != RIGHT) game.dir = LEFT;  break;
                        case 77: if (game.dir != LEFT)  game.dir = RIGHT; break;
                    }
                }
            }
            Sleep(10);
            wait += 10;
        }

        if (game.alive) tick(game);
    }

    show_cursor();
    set_cursor(0, game.height + 5);
    std::cout << std::endl;
    std::cout << "  GAME OVER!" << std::endl;
    std::cout << "  Final Score: " << game.score << std::endl;
    std::cout << "  Snake Length: " << game.snake.size() << std::endl;
    std::cout << std::endl;
    std::cout << "  Press any key to exit..." << std::endl;
    _getch();

    return 0;
}
